Add architecture documentation (#2165)
stacklok/toolhive: docs/arch/01-deployment-modes.md

# Deployment Modes

ToolHive supports three distinct deployment modes, each optimized for different use cases and environments. This document provides a detailed explanation of how ToolHive operates in each mode.

## Overview

```mermaid
graph LR
    subgraph LocalDeployment[Local Deployment]
        CLI[CLI Mode<br/>thv binary]
        UI[UI Mode<br/>ToolHive Studio]
    end

    subgraph KubernetesDeployment[Kubernetes Deployment]
        Operator[Operator Mode<br/>thv-operator]
    end

    CLI --> Docker[Docker/Podman<br/>Colima<br/>Rancher Desktop]
    UI --> Docker
    Operator --> K8s[Kubernetes]

    Docker --> MCP1[MCP Servers]
    K8s --> MCP2[MCP Servers]

    style LocalDeployment fill:#e1f5fe
    style KubernetesDeployment fill:#fff3e0
```

## Mode Comparison

| Feature | Local CLI | Local UI | Kubernetes |
|---------|-----------|----------|------------|
| **Binary** | `thv` | `thv` (API server) | `thv-operator` + `thv-proxyrunner` |

> | **Container Runtime** | Docker/Podman/Colima | Docker/Podman/Rancher | Kubernetes |

Reviewer: The Container Runtime row in this comparison table is misleading. Both Local CLI and Local UI use the same runtime detection factory and support identical container runtimes (Podman, Docker, Colima). Rancher Desktop is detected as Docker runtime type via socket path.

Consider removing this row from the table since it suggests different runtime support when they're actually the same.

Evidence: `pkg/container/factory.go` - The same factory is used for both CLI and UI modes.

    Runtime -->|creates| Container[MCP Server Container]
    Proxy -->|stdin/stdout or HTTP| Container
    Client[MCP Client] -->|HTTP/SSE/Streamable| Proxy

    style THV fill:#90caf9
    style Proxy fill:#81c784
    style Container fill:#ffb74d
```

### How It Works

1. **User executes command**: `thv run server-name`

2. **ToolHive CLI (`cmd/thv/main.go`)**:
   - Parses command-line arguments
   - Loads or creates RunConfig
   - Instantiates workloads API (`pkg/workloads/manager.go`)

3. **Workload Manager**:
   - Detects available container runtime (Podman → Colima → Docker)
   - Creates container via Runtime API
   - Spawns detached proxy process

4. **Proxy Process**:
   - Runs as independent process (via `thv restart --foreground`)
   - Attaches to container (for stdio) or forwards HTTP traffic
   - Applies middleware chain
   - Exposes local HTTP endpoint for MCP clients

5. **State Management**:
   - RunConfig saved to `~/.toolhive/state/` (or XDG equivalent)
   - PID file for process management
   - Status file for workload state tracking

### Container Runtime Selection

**Implementation**: `pkg/container/factory.go`

> The CLI automatically detects container runtimes in this order:

Reviewer: Does it make sense to include Rancher here? (No idea, asking)

**Implementation**: Uses `adrg/xdg` package for XDG Base Directory compliance.

## Local Mode: UI

### Architecture

```mermaid
graph TB
    User[User] -->|Web Browser| Studio[ToolHive Studio<br/>Web UI]
    Studio -->|REST API| APIServer[thv serve<br/>API Server]
    APIServer -->|Internal| Workloads[Workloads Manager]
    Workloads -->|Runtime API| Runtime[Container Runtime<br/>Docker/Podman/Rancher]
    Runtime -->|creates| Container[MCP Server Container]
    Container -->|managed by| Proxy[Proxy Process]
    Client[MCP Client] -->|HTTP| Proxy

    style Studio fill:#ba68c8
    style APIServer fill:#90caf9
    style Proxy fill:#81c784
    style Container fill:#ffb74d
```

### How It Works

1. **User starts UI**: ToolHive Studio application launches

2. **Studio spawns API server**: `thv serve`
   - Starts HTTP API server on configurable port (default: 8080)
   - Exposes RESTful endpoints for workload management

3. **API Server (`pkg/api/server.go`)**:
   - Handles HTTP requests from UI
   - Delegates to Workloads Manager
   - Returns JSON responses

4. **Workload Operations**:
   - Create: `POST /api/v1/workloads`

>    - List: `GET /api/v1/workloads`

Reviewer: API endpoint path is incorrect. The actual API uses `/api/v1beta/` prefix, not `/api/v1/`.

```suggestion
   - List: `GET /api/v1beta/workloads`
```

Evidence: `pkg/api/server.go:218-229`

   - Stop: `POST /api/v1/workloads/{name}/stop`

>    - Delete: `DELETE /api/v1/workloads/{name}`

Reviewer: API endpoint path is incorrect. The actual API uses `/api/v1beta/` prefix, not `/api/v1/`.

```suggestion
   - Delete: `DELETE /api/v1beta/workloads/{name}`
```

Evidence: `pkg/api/server.go:218-229`

>    - Logs: `GET /api/v1/workloads/{name}/logs`

Reviewer: API endpoint path is incorrect. The actual API uses `/api/v1beta/` prefix, not `/api/v1/`.

```suggestion
   - Logs: `GET /api/v1beta/workloads/{name}/logs`
```

Evidence: `pkg/api/server.go:218-229`

- Separation of concerns: operator manages K8s resources, proxy-runner manages MCP traffic

### Deployment Pattern

```mermaid
graph LR
    subgraph "Namespace: default"
        Deploy[Deployment<br/>proxy-runner<br/>Replicas: 1]
        SVC[Service<br/>proxy-svc]
        STS[StatefulSet<br/>mcp-server<br/>Replicas: 1]
    end

    Deploy -->|manages| STS
    SVC -->|routes to| Deploy
    Deploy -.->|watches| STS

    style Deploy fill:#90caf9
    style STS fill:#ffb74d
    style SVC fill:#81c784
```

### Custom Resource Definitions

ToolHive provides several CRDs for managing MCP servers in Kubernetes:

- **MCPServer** - Defines an MCP server deployment with container images, transports, and middleware
- **MCPRegistry** - Manages MCP server registries from Git or ConfigMap sources

For complete examples, see the [`examples/operator/mcp-servers/`](../../examples/operator/mcp-servers/) directory, which includes:
- Basic MCP server deployments with different transports (stdio, SSE, streamable-http)
- Authentication configurations (inline OIDC, ConfigMap-based, Kubernetes-native)
- Resource and pod template customizations
- Tool filtering and middleware examples

> Full CRD API documentation is available in `cmd/thv-operator/crd-api.md`.

Reviewer: It would be nice if this was a link to the docs website

> Full CRD API documentation is available in [`cmd/thv-operator/crd-api.md`](../../cmd/thv-operator/crd-api.md).

Reviewer: The referenced file `cmd/thv-operator/crd-api.md` does not exist. Consider updating to the correct path.

```suggestion
Full CRD API documentation is available in `docs/operator/crd-api.md`.
```

Evidence: File `cmd/thv-operator/crd-api.md` does not exist in the repository.

|---------|-----------|----------|------------|
| **Binary** | `thv` | `thv` (API server) | `thv-operator` + `thv-proxyrunner` |
| **Container Runtime** | Docker/Podman/Colima | Docker/Podman/Rancher | Kubernetes |
| **Process Management** | Detached processes | API-managed | Operator-managed |
| **State Storage** | Local filesystem | Local filesystem | etcd (K8s API) |
| **Scaling** | Single machine | Single machine | Cluster-wide |
| **Best For** | Developers, CLI users | UI users, beginners | Production, multi-tenant |

## Local Mode: CLI

### Architecture

```mermaid
graph TB
    User[User] -->|CLI Commands| THV[thv binary]
    THV -->|spawn detached| Proxy[Proxy Process]
    Proxy -->|Docker API| Runtime[Container Runtime<br/>Docker/Podman/Colima]
    Runtime -->|creates| Container[MCP Server Container]
    Proxy -->|stdin/stdout or HTTP| Container
    Client[MCP Client] -->|HTTP/SSE/Streamable| Proxy

    style THV fill:#90caf9
    style Proxy fill:#81c784
    style Container fill:#ffb74d
```

### How It Works

1. **User executes command**: `thv run server-name`

2. **ToolHive CLI (`cmd/thv/main.go`)**:
   - Parses command-line arguments
   - Loads or creates RunConfig

>    - Instantiates workloads API (`pkg/workloads/manager.go:38`)

Reviewer 2: Yeah I was thinking the same

```

**MCPRegistry** (`cmd/thv-operator/api/v1alpha1/mcpregistry_types.go`):
```yaml
apiVersion: mcp.stacklok.com/v1alpha1
kind: MCPRegistry
metadata:
  name: my-registry
spec:
  source:
    type: git
    git:
      url: https://github.com/example/mcp-registry
      branch: main
      path: registry.json
  syncPolicy:
    automatic: true
    interval: 1h
```

### Operator Design Decisions

See `cmd/thv-operator/DESIGN.md` for detailed decision documentation.

**Key principles:**
- Use CRD attributes for business logic affecting reconciliation
- Use PodTemplateSpec for infrastructure concerns (node selection, resources)
- Separate sync decision logic from sync execution
- Batch status updates to reduce API server load

### State Management

Unlike local mode, Kubernetes mode stores state in:
- **etcd** (via Kubernetes API)
- **ConfigMaps** for RunConfig
- **Secrets** for sensitive data (OIDC client secrets, etc.)
- **Status subresources** for workload state

No local filesystem state required.

### Scaling Considerations

**Proxy runner:**
- Typically runs with 1 replica
- Can be scaled for HA (with session affinity)

**MCP server (StatefulSet):**
- Scales independently from proxy

> - Persistent volumes attached per replica

Reviewer 2: This isn't true? No PVs are created with MCP Servers currently
